code/modules/tutorials/_tutorial.dm

```
@@ -0,0 +1,258 @@
+		return QDEL_HINT_LETMELIVE
+	return ..()
+
+/// Checks if we should perform the tutorial for the given user, and performs if so.
+/// Use `SStutorials.suggest_tutorial` instead of calling this directly.
+/datum/tutorial_manager/proc/try_perform(mob/user, list/arguments)
+	var/datum/tutorial/tutorial = new tutorial_type(user)
+	if (!tutorial.should_perform(user))
+		qdel(tutorial)
+		return
+
+	performing_ckeys[user.ckey] = TRUE
+
+	tutorial.perform(arglist(arguments))
+
+/// Checks if the user should be given this tutorial
+/datum/tutorial_manager/proc/should_run(mob/user)
+	var/ckey = user.ckey
+
+	if (isnull(ckey))
+		return FALSE
+
+	if (ckey in finished_ckeys)
+		return FALSE
+
+	if (ckey in performing_ckeys)
+		return FALSE
+
+	if (!SSdbcore.IsConnected())
+		return CONFIG_GET(flag/give_tutorials_without_db)
+
+	var/player_join_date = user.client?.player_join_date
+	if (isnull(player_join_date))
+		return FALSE
+
+	// This works because ISO-8601 is cool
+	var/grandfather_date = initial(tutorial_type.grandfather_date)
+	if (!isnull(grandfather_date) && player_join_date < grandfather_date)
+		return FALSE
+
+	return TRUE
+
+/// Marks the tutorial as completed.
+/// Call `/datum/tutorial/proc/complete()` instead.
+/datum/tutorial_manager/proc/complete(mob/user)
+	set waitfor = FALSE
+
+	ASSERT(!isnull(user.ckey))
+
+	finished_ckeys[user.ckey] = TRUE
+	performing_ckeys -= user.ckey
+
+	SSblackbox.record_feedback("tally", "tutorial_completed", 1, "[tutorial_type]")
+	log_game("[key_name(user)] completed the [tutorial_type] tutorial.")
+
+	if (SSdbcore.IsConnected())
+		INVOKE_ASYNC(src, PROC_REF(log_completion_to_database), user.ckey)
+
+/datum/tutorial_manager/proc/log_completion_to_database(ckey)
+	PRIVATE_PROC(TRUE)
+
+	var/datum/db_query/insert_tutorial_query = SSdbcore.NewQuery(
+		"INSERT IGNORE INTO [format_table_name("tutorial_completions")] (ckey, tutorial_key) VALUES (:ckey, :tutorial_key)",
```

I would suggest using `INSERT ... ON DUPLICATE KEY UPDATE x1 = y1, x2 = y2`
`INSERT IGNORE` will ignore all errors when inserting not just from a unique key, a warning is thrown on every ignored row and strict mode will try to truncate/change inserting data to work.

---

Changing to `INSERT INTO [format_table_name("tutorial_completions")] (ckey, tutorial_key) VALUES (:ckey, :tutorial_key) ON DUPLICATE KEY UPDATE tutorial_key = tutorial_key`